opal/mca/btl/iwarp/btl_iwarp.c

```diff
@@ -0,0 +1,2009 @@
+
+static inline struct ibv_cq *create_cq_compat(struct ibv_context *context,
+        int cqe, void *cq_context, struct ibv_comp_channel *channel,
+        int comp_vector)
+{
+#if OPAL_IBV_CREATE_CQ_ARGS == 3
+    return ibv_create_cq(context, cqe, channel);
+#else
+    return ibv_create_cq(context, cqe, cq_context, channel, comp_vector);
+#endif
+}
+
+static int adjust_cq(mca_btl_iwarp_device_t *device, const int cq)
+{
+    uint32_t cq_size = device->cq_size[cq];
+
+    /* make sure we don't exceed the maximum CQ size and that we
+     * don't size the queue smaller than otherwise requested
+     */
+     if(cq_size < mca_btl_iwarp_component.ib_cq_size[cq])
+        cq_size = mca_btl_iwarp_component.ib_cq_size[cq];
+
+    if(cq_size > (uint32_t)device->ib_dev_attr.max_cqe)
+        cq_size = device->ib_dev_attr.max_cqe;
+
+    if(NULL == device->ib_cq[cq]) {
+        device->ib_cq[cq] = create_cq_compat(device->ib_dev_context, cq_size,
+#if OPAL_ENABLE_PROGRESS_THREADS == 1
+                device, device->ib_channel,
+#else
+                NULL, NULL,
+#endif
+                0);
+
+        if (NULL == device->ib_cq[cq]) {
+            mca_btl_iwarp_show_init_error(__FILE__, __LINE__, "ibv_create_cq",
+                                           ibv_get_device_name(device->ib_dev));
+            return OPAL_ERROR;
+        }
+
+#if OPAL_ENABLE_PROGRESS_THREADS == 1
+        if(ibv_req_notify_cq(device->ib_cq[cq], 0)) {
+            mca_btl_iwarp_show_init_error(__FILE__, __LINE__,
+                                           "ibv_req_notify_cq",
+                                           ibv_get_device_name(device->ib_dev));
+            return OPAL_ERROR;
+        }
+
+        if (!device->progress) {
+            int rc;
+            device->progress = true;
+            if(OPAL_SUCCESS != (rc = opal_thread_start(&device->thread))) {
+                BTL_ERROR(("Unable to create progress thread, retval=%d", rc));
+                return rc;
+            }
+        }
+#endif
+    }
+#ifdef HAVE_IBV_RESIZE_CQ
+    else if (cq_size > mca_btl_iwarp_component.ib_cq_size[cq]){
+        int rc;
+        rc = ibv_resize_cq(device->ib_cq[cq], cq_size);
+        /* For ConnectX the resize CQ is not implemented and verbs returns -ENOSYS
+         * but should return ENOSYS. So it is reason for abs */
+        if(rc && ENOSYS != abs(rc) && EOPNOTSUPP != abs(rc)) {
+            BTL_ERROR(("cannot resize completion queue, error: %d", rc));
+            return OPAL_ERROR;
+        }
+    }
+#endif
+
+    return OPAL_SUCCESS;
+}
+
+
+/* In this function we check if the device supports srq limit
+   event. We create the temporary srq, post some receive buffers - in
+   order to prevent srq limit event immediately and call the
+   "ibv_modify_srq" function. If a return value of the function not
+   success => our decision that the device doesn't support this
+   capability. */
+static int check_if_device_support_modify_srq(mca_btl_iwarp_module_t *iwarp_btl)
```

Comment:
This is irrelevant if iWARP does not support SRQ.

Reply:
If you leave SRQ support in, you might want to update the ini file to indicate which model(s) of Chelsio cards support it (I think all of them are marked as point-to-point RQs right now).